subsys/net/lib/lwm2m/lwm2m_obj_firmware.c

```diff
@@ -66,41 +73,47 @@ static struct lwm2m_engine_obj_field fields[] = {
 	OBJ_FIELD_DATA(FIRMWARE_UPDATE_DELIV_METHOD_ID, R, U8)
 };
 
-static struct lwm2m_engine_obj_inst inst;
-static struct lwm2m_engine_res res[FIRMWARE_MAX_ID];
-static struct lwm2m_engine_res_inst res_inst[RESOURCE_INSTANCE_COUNT];
+static struct lwm2m_engine_obj_inst inst[MAX_INSTANCE_COUNT];
+static struct lwm2m_engine_res res[MAX_INSTANCE_COUNT][FIRMWARE_MAX_ID];
+static struct lwm2m_engine_res_inst res_inst[MAX_INSTANCE_COUNT][RESOURCE_INSTANCE_COUNT];
 
-static lwm2m_engine_set_data_cb_t write_cb;
-static lwm2m_engine_execute_cb_t update_cb;
+static lwm2m_engine_set_data_cb_t write_cb[MAX_INSTANCE_COUNT];
+static lwm2m_engine_execute_cb_t update_cb[MAX_INSTANCE_COUNT];
 
 #ifdef CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT
-extern int lwm2m_firmware_start_transfer(char *package_uri);
+extern int lwm2m_firmware_start_transfer(uint16_t obj_inst_id, char *package_uri);
 #endif
 
+uint8_t lwm2m_firmware_get_update_state_inst(uint16_t obj_inst_id)
+{
+	return update_state[obj_inst_id];
+}
+
 uint8_t lwm2m_firmware_get_update_state(void)
```

Comment:
This function should be removed as it's not needed anymore.

Reply:
It is needed for the backwards compatibility

Reply:
Then it should be kept on the header file too.